src/lib/statetable.rs

```diff
@@ -190,36 +195,30 @@ impl StateTable {
         }
         assert!(final_state.is_some());
 
-        Ok(StateTable {actions: actions,
-                       gotos: gotos,
+        Ok(StateTable {actions,
+                       used_actions,
+                       gotos,
                        nonterms_len: grm.nonterms_len(),
                        terms_len: grm.terms_len(),
-                       reduce_reduce: reduce_reduce,
-                       shift_reduce: shift_reduce,
+                       reduce_reduce,
+                       shift_reduce,
                        final_state: final_state.unwrap()})
     }
 
 
     /// Return the action for `state_idx` and `sym`, or `None` if there isn't any.
-    pub fn action(&self, state_idx: StIdx, sym: Symbol) -> Option<Action> {
-        if let Symbol::Term(term_idx) = sym {
-            let off = StateTable::actions_offset(self.terms_len, state_idx, term_idx);
-            self.actions.get(&off).map_or(None, |x| Some(*x))
-        } else {
-            None
-        }
+    pub fn action(&self, state_idx: StIdx, term_idx: TIdx) -> Option<Action> {
+        let off = StateTable::actions_offset(self.terms_len, state_idx, term_idx);
+        self.actions.get(&off).map_or(None, |x| Some(*x))
     }
 
     /// Return an iterator over the actions of `state_idx`.
-    pub fn state_actions(&self, state_idx: StIdx) -> impl Iterator<Item=Symbol> {
-        let mut syms = Vec::new();
-        for term_idx in 0..self.terms_len {
-            let off = StateTable::actions_offset(self.terms_len, state_idx, TIdx::from(term_idx));
-            if let Some(_) = self.actions.get(&off) {
-                syms.push(Symbol::Term(TIdx::from(term_idx)))
-            }
-        }
-        StateActionIterator{symbols: syms, i: 0}
+    pub fn state_actions(&self, state_idx: StIdx) -> StateActionsIterator {
+        let start_off = usize::from(state_idx) * self.terms_len as usize;
+        StateActionsIterator{stable: &self,
+                             start: start_off,
```

Comment:
Could simplify this line by calling the variable `start`

Reply:
True.